### PR TITLE
[Qudits] Support wave function/sparse simulator

### PR DESCRIPTION
**Pull request by cduck**
Part of #933.

### CODE REVIEW

File: cirq/linalg/predicates.py [hunk 1]
```diff
@@ -238,25 +242,54 @@ def slice_for_qubits_equal_to(target_qubit_axes: Sequence[int],
...
-        result[axis] = (little_endian_qureg_value >> k) & 1
+                                                  ] * (n + (num_qubits is None))
+    target_shape = tuple(qid_shape[i] for i in target_qubit_axes)
+    if big_endian_qureg_value:
+        digits = value.big_endian_int_to_digits(big_endian_qureg_value,
+                                                base=target_shape)
+    else:
+        if little_endian_qureg_value < 0:
+            little_endian_qureg_value &= ((1 << len(target_shape)) - 1)
+        digits = value.big_endian_int_to_digits(little_endian_qureg_value,
```
Comment (Strilanc) [inline cirq/linalg/predicates.py]: Extract a `little_endian_int_to_digits` method.

File: cirq/linalg/predicates.py [hunk 2]
```diff
@@ -238,25 +242,54 @@ def slice_for_qubits_equal_to(target_qubit_axes: Sequence[int],
...
-        max(target_qubit_axes) if target_qubit_axes else -1)
+        max(target_qubit_axes) + 1 if target_qubit_axes else 0)
+    if qid_shape is None:
+        qid_shape = (2,) * n
     result: List[Union[slice, int, 'ellipsis']] = [slice(None)
-                                                  ] * (n + 2 *
-                                                       (num_qubits is None))
-    for k, axis in enumerate(target_qubit_axes):
-        result[axis] = (little_endian_qureg_value >> k) & 1
+                                                  ] * (n + (num_qubits is None))
```
Comment (Strilanc) [inline cirq/linalg/predicates.py]:
```
specified_out_size = num_qubits is not None
out_size = num_qubits if specified_out_size else max(target_qubits_axes) + 1
result = [slice(None)] * out_size
if not specified_out_size:
    result.append(Ellipsis)
```

File: cirq/sim/wave_function.py [hunk 3]
```diff
@@ -386,28 +422,35 @@ def sample_state_vector(state: np.ndarray,
...
     # it. Note that we us ints here, since numpy's choice does not allow for
     # choosing from a list of tuples or list of lists.
     result = np.random.choice(len(probs), size=repetitions, p=probs)
-    # Convert to bools and rearrange to match repetition being the outer list.
-    return np.transpose([(1 & (result >> i)).astype(np.bool) for i in
-                         range(len(indices))])
+    # Convert to individual qudit measurements.
+    meas_shape = tuple(qid_shape[i] for i in indices)
+    return np.array([
+        value.big_endian_int_to_digits(result[i], base=meas_shape[::-1])[::-1]
```
Comment (Strilanc) [inline cirq/sim/wave_function.py]: Looks like this should be using a little endian method

File: cirq/linalg/predicates.py [hunk 4]
```diff
@@ -238,25 +242,54 @@ def slice_for_qubits_equal_to(target_qubit_axes: Sequence[int],
...
             desired for of the target qubits. The integer is little endian
-            w.r.t. the target quit axes, meaning the low bit of the integer
+            w.r.t. the target qubit axes, meaning the low bit of the integer
             determines the desired value of the first targeted qubit, and so
             forth with the k'th targeted qubit's value set to
             bool(qureg_value & (1 << k)).
+        big_endian_qureg_value: Same as `little_endian_qureg_value` but big
+            endian w.r.t. to target qubit axes, meaning the low bit of the
+            integer dertemines the desired value of the last target qubit, and
+            so forth.  Specify exactly of the qureg value arguments.
```
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/linalg/predicates.py]: exactly *one* of

[the remark this answers is not in the document]
Comment (cduck) [reply cirq/linalg/predicates.py]: Fixed.

File: cirq/linalg/predicates.py (hunk 2 above)
Comment (cduck) [reply cirq/linalg/predicates.py]: Cleaned up.

**Comment by cduck** [comment]
Not implementing little endian functions.  I opened an issue #2076.

@Strilanc PTAL.

**Pull request merged**